public/app/features/dimensions/editors/FileUploader.tsx

```diff
@@ -0,0 +1,97 @@
+import React, { Dispatch, SetStateAction } from 'react';
+import { FileDropzone, useTheme2, Button, DropzoneFile } from '@grafana/ui';
+import { getBackendSrv, config } from '@grafana/runtime';
+import { GrafanaTheme2 } from '@grafana/data';
+import { css } from '@emotion/css';
+import { MediaType } from '../types';
+interface Props {
+  setNewValue: Dispatch<SetStateAction<string>>;
+  setFormData: Dispatch<SetStateAction<FormData>>;
+  mediaType: MediaType;
+  setUpload: Dispatch<SetStateAction<boolean>>;
+  getRequest: (formData: FormData) => Promise<UploadResponse>;
+}
+interface UploadResponse {
+  err: boolean;
+  path: string;
+}
+export function FileDropzoneCustomChildren({ secondaryText = 'Drag and drop here or browse' }) {
+  const theme = useTheme2();
+  const styles = getStyles(theme);
+
+  return (
+    <div className={styles.iconWrapper}>
+      <small className={styles.small}>{secondaryText}</small>
+      <Button icon="upload">Upload</Button>
+    </div>
+  );
+}
+export const FileUploader = ({ mediaType, setNewValue, setFormData, setUpload, getRequest }: Props) => {
+  const onFileRemove = (file: DropzoneFile) => {
+    fetch(`/api/storage/delete/upload/${file.file.name}`, {
+      method: 'DELETE',
+    }).catch((error) => console.error('cannot delete file', error));
+  };
+  const acceptableFiles =
+    mediaType === 'icon' ? 'image/svg+xml' : 'image/jpeg,image/png,image/gif,image/png, image/webp';
+  return (
+    <FileDropzone
+      readAs="readAsBinaryString"
+      onFileRemove={onFileRemove}
+      options={{
+        accept: acceptableFiles,
+        multiple: false,
+        onDrop: (acceptedFiles: File[]) => {
+          let formData = new FormData();
+          formData.append('file', acceptedFiles[0]);
+          getRequest(formData).then((data) => {
+            if (!data.err) {
+              getBackendSrv()
+                .get(`api/storage/read/${data.path}`)
+                .then(() => {
+                  setNewValue(`${config.appUrl}api/storage/read/${data.path}`);
+                });
+            }
+          });
```

Comment:
I find this approach a bit confusing, we can customise the upload function but not the callback. In this case it is assumed that custom `getRequest` will use the same `api/storage` root as the callback. So I'm wondering why not either move `getRequest` logic into this component or make the whole `onDrop` function a prop?

Reply:
agree - I ended up moving all the network calls to the parent component instead of passing this getRequest to FileUploader.